### ISSUE TITLE
postgres DB migrated from an (old) sqlite fails with lots of duplicate state_groups_pkey violations

### ISSUE THREAD

**Issue by ara4n** [matrix-org/synapse#3050]
I took an old 2016-vintage homeserver.db; ran it with synapse 0.27; then migrated it to postgres via the  migration script.  All seemed to work.

However, on using it, my logs have been filling up with lots of:

```
2018-03-29 14:47:07,652 - synapse.federation.federation_server - 169 - ERROR - PUT-44876 - Failed to handle PDU $1522331212167qlKlG:ldbco.de
Traceback (most recent call last):
  File "synapse/federation/federation_server.py", line 161, in process_pdus_for_room
    transaction.origin, pdu
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/federation/federation_server.py", line 539, in _handle_received_pdu
    yield self.handler.on_receive_pdu(origin, pdu, get_missing=True)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/handlers/federation.py", line 238, in on_receive_pdu
    auth_chain=auth_chain,
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/handlers/federation.py", line 347, in _process_received_pdu
    origin, auth_chain, state, event
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/handlers/federation.py", line 1591, in _persist_auth_tree
    event, old_state=state
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/state.py", line 251, in compute_event_context
    current_state_ids=context.current_state_ids,
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/storage/_base.py", line 320, in runInteraction
    result = yield self.runWithConnection(inner_func, *args, **kwargs)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/internet/defer.py", line 1384, in _inlineCallbacks
    result = result.throwExceptionIntoGenerator(g)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/failure.py", line 408, in throwExceptionIntoGenerator
    return g.throw(self.type, self.value, self.tb)
  File "synapse/storage/_base.py", line 365, in runWithConnection
    inner_func, *args, **kwargs
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
    result = inContext.theWork()
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
    inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
    return func(*args,**kw)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 307, in _runWithConnection
    compat.reraise(excValue, excTraceback)
  File "/home/matrix/.synapse/local/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 298, in _runWithConnection
    result = func(conn, *args, **kw)
  File "synapse/storage/_base.py", line 361, in inner_func
    return func(conn, *args, **kwargs)
  File "synapse/storage/_base.py", line 316, in inner_func
    func, *args, **kwargs
  File "synapse/storage/_base.py", line 241, in _new_transaction
    r = func(txn, *args, **kwargs)
  File "synapse/storage/state.py", line 613, in _store_state_group_txn
    "event_id": event_id,
  File "synapse/storage/_base.py", line 444, in _simple_insert_txn
    txn.execute(sql, vals)
  File "synapse/storage/_base.py", line 82, in execute
    self._do_execute(self.txn.execute, sql, *args)
  File "synapse/storage/_base.py", line 112, in _do_execute
    sql, *args
IntegrityError: duplicate key value violates unique constraint "state_groups_pkey"
DETAIL:  Key (id)=(86) already exists.
```

...which should surely never happen?

**Comment by ara4n** [comment]
@erikjohnston just highlighting this in case it impacts on your state calc machinations atm.

**Comment by ara4n** [comment]
(is it possible that it was still doing some background schema updates when I killed it and preemptively migrated it to postgres whilst it was in an inconsistent state?)

**Comment by ZerataX** [comment]
I'm running synapse 0.27.2-1 installed from the debian package on ubuntu 16.04.4 and I'm getting similar errors from a few days old sqlite db ported to postgres.

**Comment by richvdh** [comment]
Oh boo. Looks like we aren't correctly populating the `state_group_id_seq` which was introduced in #2784 / 0.27.0.

Easiest fix is going to be (with synapse stopped):

```
SELECT MAX(id)+1 AS next_id FROM state_groups \gset
ALTER SEQUENCE state_group_id_seq RESTART WITH :next_id;
```

**Comment by richvdh** [comment]
Hopefully fixed as of 0.27.3.
